File: app/src/main/java/com/hydbest/baseandroid/activity/plugin/ApkLoaderActivity.java
package com.hydbest.baseandroid.activity.plugin;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Environment;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;

import com.hydbest.baseandroid.R;

import java.io.File;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import dalvik.system.DexClassLoader;
import dalvik.system.PathClassLoader;

/**
 * Created by csz on 2019/2/1.
 */

public class ApkLoaderActivity extends AppCompatActivity {

    @BindView(R.id.iv)
    ImageView mIv;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_apk_load);
        ButterKnife.bind(this);
    }

    /**
     * 找出所有已安装插件
     *
     * @return
     */
    private List<PluginBean> findAllPlugin() {
        List<PluginBean> list = new ArrayList<>();
        PackageManager pm = getPackageManager();
        List<PackageInfo> installedPackages = pm.getInstalledPackages(PackageManager.GET_UNINSTALLED_PACKAGES);
        for (PackageInfo packageInfo : installedPackages) {
            String packName = packageInfo.packageName;
            String shareUserId = packageInfo.sharedUserId;
            if (shareUserId != null && shareUserId.equals("share") && !"com.hydbest.baseandroid".equals(packName)) {
                String label = pm.getApplicationLabel(packageInfo.applicationInfo).toString();
                Log.i("csz", "label:" + label);
                list.add(new PluginBean(label, packName));
            }
        }
        return list;
    }

    private int dynamicLoadResouce(String packName) throws Exception {
        Context pluginContext = createPackageContext(packName, CONTEXT_IGNORE_SECURITY | CONTEXT_INCLUDE_CODE);
        Log.i("csz", "资源路径：" + pluginContext.getPackageResourcePath());
        PathClassLoader pathClassLoader = new PathClassLoader(pluginContext.getPackageResourcePath(), ClassLoader.getSystemClassLoader());
        Class clazz = Class.forName(packName + ".R$mipmap", true, pathClassLoader);
        Field field = clazz.getDeclaredField("one");
        int resourceId = field.getInt(null);
        Log.i("csz", "资源Id:" + resourceId);
        Drawable drawable = pluginContext.getResources().getDrawable(resourceId);

        return resourceId;
    }

    public void loadPlugin(View view) {
        dynamicLoadApk("com.csz.ni", Environment.getExternalStorageDirectory() + File.separator + "plugin.apk");
    }

    /**
     * 获取插件包名等信息
     */
    private String[] getUninstallApk(String uninstallApkPath) {
        String[] info = new String[2];
        PackageManager pm = getPackageManager();
        PackageInfo packInfo = pm.getPackageArchiveInfo(uninstallApkPath, PackageManager.GET_ACTIVITIES);
        if (packInfo != null) {
            ApplicationInfo applicationInfo = packInfo.applicationInfo;
            String label = pm.getApplicationLabel(applicationInfo).toString();
            String packageName = packInfo.packageName;
            info[0] = label;
            info[1] = packageName;
        }
        return info;
    }

    private Resources getPluginResources(String pluginPath) {
        try {
            AssetManager assetManager = AssetManager.class.newInstance();
            //getDeclaredMethod
            Method addAssetPath = assetManager.getClass().getDeclaredMethod("addAssetPath", String.class);
            addAssetPath.invoke(assetManager, pluginPath);
            Resources origin = getResources();
            Resources plugin = new Resources(assetManager, origin.getDisplayMetrics(), origin.getConfiguration());
            plugin.newTheme().setTo(getTheme());
            return plugin;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 加载外部资源
     *
     * @param packageName
     * @param pluginPath
     */
    private void dynamicLoadApk(String packageName, String pluginPath) {
        try {
            File dexPath = getDir("dex", Context.MODE_PRIVATE);
            Log.i("csz", "dexPath:" + dexPath);
            DexClassLoader dexClassLoader = new DexClassLoader(pluginPath, dexPath.getAbsolutePath(), null, ClassLoader.getSystemClassLoader());
            Class<?> aClass = dexClassLoader.loadClass(packageName + ".R$mipmap");
            Field one = aClass.getDeclaredField("one");
            int resourceId = one.getInt(null);
            Resources pluginResources = getPluginResources(pluginPath);
            if (pluginResources != null) {
                Drawable bg = pluginResources.getDrawable(resourceId);
                mIv.setImageDrawable(bg);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    class PluginBean {
        private String label;
        private String packName;

        public PluginBean(String label, String packName) {
            this.label = label;
            this.packName = packName;
        }
    }

}

File: app/src/main/java/com/hydbest/baseandroid/activity/plugin/ApkLoaderActivity.kt
package com.hydbest.baseandroid.activity.plugin

import android.content.Context
import android.content.pm.PackageManager
import android.content.res.AssetManager
import android.content.res.Resources
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.hydbest.baseandroid.R
import dalvik.system.DexClassLoader
import dalvik.system.PathClassLoader
import kotlinx.android.synthetic.main.activity_apk_load.*
import java.io.File
import java.util.*

/**
 * Created by csz on 2019/2/1.
 */
class ApkLoaderActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_apk_load)
    }

    /**
     * 找出所有已安装插件
     *
     * @return
     */
    private fun findAllPlugin(): List<PluginBean> {
        val list: MutableList<PluginBean> = ArrayList()
        val pm = packageManager
        val installedPackages = pm.getInstalledPackages(PackageManager.GET_UNINSTALLED_PACKAGES)
        for (packageInfo in installedPackages) {
            val packName = packageInfo.packageName
            val shareUserId = packageInfo.sharedUserId
            if (shareUserId != null && shareUserId == "share" && "com.hydbest.baseandroid" != packName) {
                val label = pm.getApplicationLabel(packageInfo.applicationInfo).toString()
                Log.i("csz", "label:$label")
                list.add(PluginBean(label, packName))
            }
        }
        return list
    }

    @Throws(Exception::class)
    private fun dynamicLoadResouce(packName: String): Int {
        val pluginContext = createPackageContext(packName, Context.CONTEXT_IGNORE_SECURITY or Context.CONTEXT_INCLUDE_CODE)
        Log.i("csz", "资源路径：" + pluginContext.packageResourcePath)
        val pathClassLoader = PathClassLoader(pluginContext.packageResourcePath, ClassLoader.getSystemClassLoader())
        val clazz = Class.forName("$packName.R\$mipmap", true, pathClassLoader)
        val field = clazz.getDeclaredField("one")
        val resourceId = field.getInt(null)
        Log.i("csz", "资源Id:$resourceId")
        val drawable = pluginContext.resources.getDrawable(resourceId)
        return resourceId
    }

    fun loadPlugin(view: View?) {
        dynamicLoadApk("com.csz.ni", Environment.getExternalStorageDirectory().toString() + File.separator + "plugin.apk")
    }

    /**
     * 获取插件包名等信息
     */
    private fun getUninstallApk(uninstallApkPath: String): Array<String?> {
        val info = arrayOfNulls<String>(2)
        val pm = packageManager
        val packInfo = pm.getPackageArchiveInfo(uninstallApkPath, PackageManager.GET_ACTIVITIES)
        if (packInfo != null) {
            val applicationInfo = packInfo.applicationInfo
            val label = pm.getApplicationLabel(applicationInfo).toString()
            val packageName = packInfo.packageName
            info[0] = label
            info[1] = packageName
        }
        return info
    }

    private fun getPluginResources(pluginPath: String): Resources? {
        try {
            val assetManager = AssetManager::class.java.newInstance()
            //getDeclaredMethod
            val addAssetPath = assetManager.javaClass.getDeclaredMethod("addAssetPath", String::class.java)
            addAssetPath.invoke(assetManager, pluginPath)
            val origin = resources
            val plugin = Resources(assetManager, origin.displayMetrics, origin.configuration)
            plugin.newTheme().setTo(theme)
            return plugin
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return null
    }

    /**
     * 加载外部资源
     *
     * @param packageName
     * @param pluginPath
     */
    private fun dynamicLoadApk(packageName: String, pluginPath: String) {
        try {
            val dexPath = getDir("dex", Context.MODE_PRIVATE)
            Log.i("csz", "dexPath:$dexPath")
            val dexClassLoader = DexClassLoader(pluginPath, dexPath.absolutePath, null, ClassLoader.getSystemClassLoader())
            val aClass = dexClassLoader.loadClass("$packageName.R\$mipmap")
            val one = aClass.getDeclaredField("one")
            val resourceId = one.getInt(null)
            val pluginResources = getPluginResources(pluginPath)
            if (pluginResources != null) {
                val bg = pluginResources.getDrawable(resourceId)
                iv.setImageDrawable(bg)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    internal inner class PluginBean(private val label: String, private val packName: String)
}